frontend_tests/zjsunit/namespace.js

```diff
@@ -31,7 +31,25 @@ function require_path(name, fn) {
 exports.zrequire = function (name, fn) {
     fn = require_path(name, fn);
     requires.push(fn);
-    return require(fn);
+
+    // The proxy below only effects modules that are not
+    // in window namespace, whose zrequire return
+    // value is used with exception of node_module zrequire.
+    return new Proxy(require(fn), {
+        set(target, prop, value) {
+            // We use rewire and set the value directly on to
+            // the module. Rewire only work for modules that
+            // are converted to ES6, or TS. Directly setting the
+            // value is required for commonjs modules that use exports.
+            // Doing both is harmless and has no side-effects.
+            target[prop] = value;
+            if ("__Rewire__" in target) {
+                target.__Rewire__(prop, value);
+            }
```

Comment:
This seems too magical for what minor syntactic sugar benefit it provides. Reaching into a module to edit a read-only value is a weird operation that *should* look weird; our tests shouldn’t just silently allow it. Additionally, it would be nice to eventually get rid of `zrequire` altogether.

How much harder is it to just write the `__Rewire__` calls where we need them?

Reply:
Not sure, I can try it out though. I did this under the assumption that we don't want other to learn new rewire stuff for writing test that they otherwise would have able to do without it previously.

Reply:
> Reaching into a module to edit a read-only value is a weird operation that should look weird

Unless I am missing something, I think we would need to use `__Rewire__` the most for mocking function in which this doesn't apply?

Reply:
These are the [only changes](https://github.com/priyank-p/zulip/commit/6e56cb983fad730cf7a258a0413bd9842c63f428) needed if the `proxy` commit is removed. Updating this PR using `__Rewire__` calls.